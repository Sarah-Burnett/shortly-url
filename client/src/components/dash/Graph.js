import React, { useState, useEffect } from "react";
import MediaQuery from "react-responsive";
import { BarChart, XAxis, YAxis, Bar } from "recharts";
import styled from "styled-components";
import { darkViolet, cyan, white, grey, red, breakpoint } from "../Styles";
import Error from "../Error";
import { useActiveLink } from "../../hooks/useActiveLink";

const GraphBox = styled.div`
    background: ${darkViolet};
    min-height: 50vh;
    padding: 2vh 2vw;
    color: ${grey};
    .date span {
            font-size: larger;
            color: ${cyan};
            white-space: nowrap;
            @media (min-width: ${breakpoint}){
                display: block;
            }
        }
    } 
    .error {
        font-size: smaller;
    }
    @media (min-width: ${breakpoint}) {
        display: flex;
        align-items: flex-start;
        justify-content: flex-start;
        div {
            height: 50vh;
            display: flex;
            flex-direction: column;
            justify-content: space-between;
        }
    }
`;

const DateInput = styled.input`
	background: none;
	color: ${white};
	width: auto;
	display: block;
	padding: 1vh 1vw;
	margin: 1vh 0;
	border-radius: 5px;
	border: 1px solid ${grey};
	:invalid {
		border: 1px solid ${red};
		outline: 1px solid ${red};
	}
`;

const Graph = () => {
	const { activeLink } = useActiveLink();
	const [firstDate, setFirstDate] = useState(Date.now() - 518400000);
	const [clickData, setClickData] = useState([]);
	const [totalClicks, setTotalClicks] = useState(0);
	const [error, setError] = useState("");
	const styles = {
		fontSize: "smaller",
	};

	useEffect(() => {
		const data = [];
		const clickArr = [[], [], [], [], [], [], []];
		const dateArr = [new Date(firstDate).toString().substring(0, 10)];
		for (let i = 1; i < 7; i++) {
			dateArr.push(
				new Date(i * 86400000 + firstDate).toString().substring(0, 10)
			);
		}
		activeLink.click
			.filter((click) =>
				dateArr.includes(new Date(click.date).toString().substring(0, 10))
			)
			.map((click) => {
				const index = dateArr.indexOf(
					new Date(click.date).toString().substring(0, 10)
				);
				clickArr[index].push(new Date(click.date));
			});
		for (let i = 0; i < 7; i++) {
			data[i] = {
				date: dateArr[i].substring(4, 10),
				clicks: clickArr[i].length,
			};
		}
		setClickData(data);
	}, [firstDate, activeLink]);

	useEffect(() => {
		setTotalClicks(() =>
			clickData.reduce((sum, date) => (sum += date.clicks), 0)
		);
	}, [clickData]);

	const selectDate = (event) => {
		setError("");
		event.target.validity.valid
			? setFirstDate(Math.floor(new Date(event.target.value).getTime()))
			: setError(["Please enter date in format 'yyyy-mm-dd'"]);
	};

	return (
		<GraphBox>
			<div className="date">
				<DateInput type="date" id="date" onChange={selectDate} readonly />
				{error !== "" && <Error error={error} />}
				<p>
					WEEK COMMENCING:{" "}
					<span>{new Date(firstDate).toString().substring(0, 10)}</span>
				</p>
				<p>
					WEEK COUNT: <span>{totalClicks}</span>
				</p>
			</div>
			<MediaQuery maxWidth={breakpoint}>
				<BarChart width={300} height={300} data={clickData} style={styles}>
					<XAxis dataKey="date" />
					<YAxis />
					<Bar dataKey="clicks" barSize={20} fill={cyan} />
				</BarChart>
			</MediaQuery>
			<MediaQuery minWidth={breakpoint}>
				<BarChart width={600} height={300} data={clickData} style={styles}>
					<XAxis dataKey="date" />
					<YAxis />
					<Bar dataKey="clicks" barSize={30} fill={cyan} />
				</BarChart>
			</MediaQuery>
		</GraphBox>
	);
};

export default Graph;
