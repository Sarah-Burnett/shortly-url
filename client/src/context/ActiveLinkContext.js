import React, { useState, createContext } from 'react';

export const ActiveLinkContext = createContext();

export const ActiveLinkProvider = (props) => {
    const [ activeLink, setActiveLink] = useState({
        activeLink: [],
        process: false
    });

    return (
        <ActiveLinkContext.Provider value={[activeLink, setActiveLink]}>
            {props.children}
        </ActiveLinkContext.Provider>
    )
};