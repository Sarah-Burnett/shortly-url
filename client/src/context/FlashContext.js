import React, { useState, createContext } from 'react';

export const FlashContext = createContext();

export const FlashProvider = (props) => {
    const [ flash, setFlash ] = useState({
        flashMsg: "",
        isFlashVisible: false,
        isAutoHideFlash: true,
    });
    return (
        <FlashContext.Provider value={[flash, setFlash]}>
            {props.children}
        </FlashContext.Provider>
    )
};