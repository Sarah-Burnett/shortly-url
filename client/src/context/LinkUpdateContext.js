import React, { useState, createContext } from 'react';

export const LinkUpdateContext = createContext();

export const LinkUpdateProvider = (props) => {
    const [ linkUpdate, setLinkUpdate] = useState();
    return (
        <LinkUpdateContext.Provider value={[linkUpdate, setLinkUpdate]}>
            {props.children}
        </LinkUpdateContext.Provider>
    )
};