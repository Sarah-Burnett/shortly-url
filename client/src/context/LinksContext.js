import React, { useState, createContext } from 'react';

export const LinksContext = createContext();

export const LinksProvider = (props) => {
    const [ links, setLinks] = useState([]);

    return (
        <LinksContext.Provider value={[links, setLinks]}>
            {props.children}
        </LinksContext.Provider>
    )
};