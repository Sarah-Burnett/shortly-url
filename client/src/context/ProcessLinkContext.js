import React, { useState, createContext } from 'react';

export const ProcessLinkContext = createContext();

export const ProcessLinkProvider = (props) => {
    const [ processLink, setProcessLink] = useState({
        link: [],
        process: false
    });
    return (
        <ProcessLinkContext.Provider value={[processLink, setProcessLink]}>
            {props.children}
        </ProcessLinkContext.Provider>
    )
};