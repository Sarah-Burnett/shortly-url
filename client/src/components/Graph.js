import React, { useState, useEffect, useContext } from 'react';
import MediaQuery from 'react-responsive';
import { BarChart, XAxis, YAxis, Bar } from 'recharts';
import { ActiveLinkContext } from '../context/ActiveLinkContext';
import styled from 'styled-components';
import { darkViolet, cyan, white, grey, red, breakpoint, Input } from '../Styles';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faExclamationTriangle } from '@fortawesome/free-solid-svg-icons';

const GraphBox = styled.div`
    background: ${darkViolet};
    min-height: 50vh;
    padding: 2vh 2vw;
    color: ${grey};
    .date span {
            font-size: larger;
            color: ${cyan};
            white-space: nowrap;
            @media (min-width: ${breakpoint}){
                display: block;
            }
        }
    }
    .error {
        font-size: smaller;
    }
    @media (min-width: ${breakpoint}) {
        display: flex;
        align-items: flex-start;
        justify-content: flex-start;
        div {
            height: 50vh;
            display: flex;
            flex-direction: column;
            justify-content: space-between;
        }
    }
`

const DateInput = styled(Input)`
    background: none;
    color: ${white};
    width: auto;
    display: block;
`

const Error = styled.p`
    color: ${red};
    border: 1px solid ${red};
    padding: 1vh 1vw;
    border-radius: 5px;
`


export const Graph = ({loading}) => {
    const [ activeLink, setActiveLink ] = useContext(ActiveLinkContext);
    const [ firstDate, setFirstDate ] = useState(Date.now() - 518400000);
    const [ clickData, setClickData ] = useState([]);
    const [ totalClicks, setTotalClicks ] = useState(0);
    const [ error, setError ] = useState('');

    const styles = {
        fontSize: "smaller",
    }

    useEffect( () => {
        const data = [];
        const clickArr = [[], [], [], [], [], [], []];
        const dateArr = [new Date(firstDate).toString().substring(0,10)];
        for (let i = 1; i < 7; i++){
            dateArr.push(new Date((i * 86400000) + firstDate).toString().substring(0,10));
        }
        activeLink.click
            .filter( click => dateArr.includes(new Date(click.date).toString().substring(0, 10)))
            .map( click => {
                const index = dateArr.indexOf(new Date(click.date).toString().substring(0, 10));
                clickArr[index].push(new Date(click.date)); 
            })
        for (let i = 0; i < 7; i++){
            data[i] = {date: dateArr[i].substring(4, 10), clicks: clickArr[i].length}
        }
        setClickData(data);
    }, [firstDate, activeLink]);

    useEffect( () => {
        setTotalClicks( () => clickData.reduce( (sum, date) => sum += date.clicks, 0 ));
    }, [clickData])
    
    const selectDate = (event) => {
        setError('');
        event.target.validity.valid ? setFirstDate(Math.floor(new Date(event.target.value).getTime())) : setError(["Please enter date in format 'yyyy-mm-dd'"]);
    }

    return (
        <GraphBox>
            <div className="date">
                <DateInput type="date" id="date" onChange={selectDate} readonly/>
                { error !== '' && <Error className="error"><FontAwesomeIcon icon={faExclamationTriangle}/> {error}</Error> }
                <p>WEEK COMMENCING: <span>{new Date(firstDate).toString().substring(0, 10)}</span></p>
                <p>WEEK COUNT: <span>{totalClicks}</span></p>
            </div>
            <MediaQuery maxWidth={breakpoint}>
                <BarChart width={300} height={300} data={clickData} style={styles}>
                    <XAxis dataKey="date"/>
                    <YAxis/>
                    <Bar dataKey="clicks" barSize={20} fill={cyan}/>
                </BarChart>
            </MediaQuery>
            <MediaQuery minWidth={breakpoint}>
                <BarChart width={600} height={300} data={clickData} style={styles}>
                    <XAxis dataKey="date"/>
                    <YAxis/>
                    <Bar dataKey="clicks" barSize={30} fill={cyan}/>
                </BarChart>
            </MediaQuery>
        </GraphBox>
    )
}