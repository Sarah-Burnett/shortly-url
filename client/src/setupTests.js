import React from 'react';
import ReactDOM from 'react-dom';
import '@testing-library/jest-dom/extend-expect';
import { render, cleanup, fireEvent, screen, waitForElement } from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import renderer from 'react-test-renderer';
import { UserProvider } from './context/UserContext';
import { LinksProvider } from './context/LinksContext';
import { ActiveLinkProvider } from './context/ActiveLinkContext';
import { ProcessLinkProvider } from './context/ProcessLinkContext';
import { FlashProvider } from './context/FlashContext';
import App from './App';

export const ComponentWrapper = (Component) => {
    return (
        <UserProvider>
            <LinksProvider>
                <ActiveLinkProvider>
                    <ProcessLinkProvider>
                        <FlashProvider>
                            <Component/>
                        </FlashProvider>
                    </ProcessLinkProvider>
                </ActiveLinkProvider>
            </LinksProvider>
        </UserProvider>
    )
}

export const rendersWithoutCrashing = (Component) => {
    const div = document.createElement('div');
    ReactDOM.render(<Component/>, div);
}

export const matchesSnapshot = (Component) => {
    const tree = renderer.create(<Component/>).toJSON();
    expect(tree).toMatchSnapshot();
}

export const sampleLinks = [
    {
        "date":"2020-08-14T17:29:53.041Z",
        "click":[
            {"date":1597426302926},
            {"date":1597426306071},
            {"date":1597426310231},
            {"date":1597426314763},
            {"date":1597428206995},
            {"date":1597428210093},
            {"date":1597428215131},
            {"date":1597428218755}
        ],
        "_id":"K1pjMCK-F",
        "full":"http://reactcommunity.org/react-transition-group/css-transition",
        "userId":"5ebe7d235e4a501ee8699058",
        "short":"octopus","__v":8
    },
    {
        "date":"2020-08-14T17:33:55.748Z",
        "click":[],
        "_id":"lsncCe5RV",
        "full":"https://ba-grant.netlify.app/",
        "userId":"5ebe7d235e4a501ee8699058",
        "short":"bagrant","__v":0
    }
];
