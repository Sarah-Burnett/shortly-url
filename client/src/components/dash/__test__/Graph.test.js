import React, { useState } from 'react';
import Graph from '../Graph';
import { cleanup } from '@testing-library/react';
import { ActiveLinkContext } from '../../../context/ActiveLinkContext';
import { rendersWithoutCrashing, matchesSnapshot, sampleLinks } from '../../../setupTests';

afterEach(cleanup);

const GraphWrapper = () => {
    const sampleLink = sampleLinks[0];
    const [ activeLink, setActiveLink] = useState(sampleLink);
    return (
        <ActiveLinkContext.Provider value={[activeLink, setActiveLink]}>
            <Graph/>
        </ActiveLinkContext.Provider>
    )
}

it('renders graph without crashing', () => rendersWithoutCrashing(GraphWrapper));

it('graph matches snapshot', () => matchesSnapshot(GraphWrapper));

//TODO: test graph functioning
