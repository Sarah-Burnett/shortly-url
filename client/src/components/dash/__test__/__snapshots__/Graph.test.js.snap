// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graph matches snapshot 1`] = `
<div
  className="sc-fznyAO isDytu"
>
  <div
    className="date"
  >
    <input
      className="sc-fznKkj lnSHCc"
      id="date"
      onChange={[Function]}
      type="date"
    />
    <p>
      WEEK COMMENCING:
       
      <span>
        Tue Aug 18
      </span>
    </p>
    <p>
      WEEK COUNT: 
      <span>
        0
      </span>
    </p>
  </div>
</div>
`;
